File: backend-golang-rest/pkg/tasks/db_in_memory.go
package tasks

import (
	"sync"

	"github.com/tonitienda/kadai/backend-golang-rest/pkg/common"
)

type InMemoryTasksDB struct {
	tasksByID    map[string]Task
	tasksByOwner map[string][]Task
	lock         sync.RWMutex
}

func NewInMemoryTasksDB() *InMemoryTasksDB {
	return &InMemoryTasksDB{
		tasksByID:    make(map[string]Task),
		tasksByOwner: make(map[string][]Task),
	}
}

func (db *InMemoryTasksDB) GetTasks(ownerID string) ([]Task, error) {
	db.lock.RLock()
	tasks, ok := db.tasksByOwner[ownerID]
	db.lock.RUnlock()

	if !ok {
		return []Task{}, nil
	}

	return tasks, nil
}

func (db *InMemoryTasksDB) GetTask(taskID string) (Task, bool) {
	db.lock.RLock()
	task, ok := db.tasksByID[taskID]
	db.lock.RUnlock()

	return task, ok
}

func (db *InMemoryTasksDB) AddTask(task Task) error {
	db.lock.Lock()
	db.tasksByID[task.ID] = task
	db.tasksByOwner[task.OwnerID] = append(db.tasksByOwner[task.OwnerID], task)
	db.lock.Unlock()
	return nil
}

func (db *InMemoryTasksDB) DeleteTask(id string) error {
	db.lock.Lock()
	task := db.tasksByID[id]
	delete(db.tasksByID, id)

	tasks := db.tasksByOwner[task.OwnerID]

	taskIndex := -1
	for i, t := range tasks {
		if t.ID == id {
			taskIndex = i
			break
		}
	}

	if taskIndex == -1 {
		db.lock.Unlock()
		return common.NotFoundError{}
	}

	db.tasksByOwner[task.OwnerID] = append(tasks[:taskIndex], tasks[taskIndex+1:]...)
	db.lock.Unlock()
	return nil
}

File: backend-rest-js/src/tasks/db-in-memory.js
const TasksByOwner = {};
const TasksById = {};

async function getTasks(ownerId) {
  console.log("Getting tasks");

  if (!TasksByOwner[ownerId]) {
    TasksByOwner[ownerId] = [];
  }

  const nonDeletedTasks = TasksByOwner[ownerId].filter((t) => !t.deletedAt);

  console.log("nonDeletedTasks", nonDeletedTasks);

  return nonDeletedTasks;
}

function getTaskById(taskId) {
  return Promise.resolve(TasksById[taskId]);
}

function addTask(task) {
  console.log("Inserting", task);

  if (!TasksByOwner[task.ownerId]) {
    TasksByOwner[task.ownerId] = [];
  }

  TasksByOwner[task.ownerId].push(task);
  TasksById[task.id] = task;
}

function updateTask(task) {
  console.log("Updating", task);
  if (!TasksByOwner[task.ownerId]) {
    TasksByOwner[task.ownerId] = [];
  }

  TasksById[task.id] = task;
  for (let idx = 0; idx < TasksByOwner[task.ownerId].length; idx++) {
    if (TasksByOwner[task.ownerId][idx].id == task.id) {
      TasksByOwner[task.ownerId][idx] = task;
      return;
    }
  }
}

module.exports = {
  getTasks,
  getTaskById,
  addTask,
  updateTask,
};
